api/src/sparql.rs

```diff
@@ -0,0 +1,102 @@
+//! Common traits for working with [SPARQL](https://www.w3.org/TR/sparql11-query/).
+//!
+//! # Design rationale
+//!
+//! These traits are deliberately very generic.
+//! Specific implementations may have additional features, such as:
+//!
+//! - preparing a query for multiple use;
+//! - setting default values for `BASE`, `PREFIX`, `FROM`, `FROM NAMED` directives,
+//!   before parsing query string;
+//! - pre-binding variables before evaluating query;
+//! - etc...
+//!
+//! However, we do not want to impose these feature, or any subset thereof,
+//! to all implementation of Sophia.
+
+use crate::graph::Graph;
+use crate::term::TTerm;
+use std::error::Error;
+
+/// A dataset that can be queried with SPARQL.
+pub trait SparqlDataset {
+    type BindingsTerm: TTerm;
+    type BindingsResult: SparqlBindings<Self>;
+    type GraphResult: Graph;
```

Comment:
Returning a `Graph` requires to collect all the results and not return them in a streaming way. We could maybe use an iterator instead?

Reply:
Good idea, a TripleSource would be more appropriate.